File: spikes/telegram-engine/adapters/telethon/behavior_resolve_targets.py
"""Resolve controlled Telegram targets without join/send side effects."""

from __future__ import annotations

import argparse
import asyncio
import json
import os
import platform
import sys
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from adapter import SessionConfig, TelethonAdapter, create_telethon_adapter


SCENARIO = "resolve_controlled_targets"
CANDIDATE = "telethon"
ADAPTER_VERSION = "Telethon==1.44.0"
TARGET_ROLES = {
    "public_group": "TELEGRAM_TEST_PUBLIC_TARGET",
    "approval_group": "TELEGRAM_TEST_APPROVAL_TARGET",
    "discussion_channel": "TELEGRAM_TEST_DISCUSSION_CHANNEL",
}


@dataclass(frozen=True)
class RunResult:
    records: list[dict[str, Any]]
    passed: bool


def metadata() -> dict[str, Any]:
    return {
        "type": "metadata",
        "candidate": CANDIDATE,
        "runtime": "python",
        "runtimeVersion": platform.python_version(),
        "adapterVersion": ADAPTER_VERSION,
        "scenarioSet": "behavior-resolve-v1",
    }


def failure_code(error: Exception) -> str:
    code = getattr(error, "code", None)
    return code if isinstance(code, str) and code else "TARGET_RESOLVE_FAILED"


async def run_resolve_targets(
    create_adapter: Callable[[], TelethonAdapter],
    targets: dict[str, str],
    *,
    clock_ns: Callable[[], int] = time.perf_counter_ns,
) -> RunResult:
    records: list[dict[str, Any]] = [metadata()]
    adapter = create_adapter()
    try:
        await adapter.connect()
        for role in TARGET_ROLES:
            target = targets.get(role, "").strip()
            if not target:
                raise RuntimeError(f"{TARGET_ROLES[role]} wajib diisi untuk behavior test.")
            started = clock_ns()
            resolved = await adapter.resolve_target(target)
            elapsed_ms = max(0, clock_ns() - started) / 1_000_000
            records.append(
                {
                    "type": "sample",
                    "candidate": CANDIDATE,
                    "scenario": SCENARIO,
                    "sessions": 1,
                    "metric": "resolve_latency",
                    "value": elapsed_ms,
                    "unit": "ms",
                    "targetRole": role,
                    "entityType": resolved["entityType"],
                }
            )
            records.append(
                {
                    "type": "assertion",
                    "candidate": CANDIDATE,
                    "scenario": SCENARIO,
                    "name": f"{role}_resolved",
                    "passed": True,
                    "hardGate": True,
                    "targetRole": role,
                    "entityType": resolved["entityType"],
                }
            )
    except Exception as error:
        records.append(
            {
                "type": "assertion",
                "candidate": CANDIDATE,
                "scenario": SCENARIO,
                "name": "all_targets_resolved",
                "passed": False,
                "hardGate": True,
                "code": failure_code(error),
            }
        )
        return RunResult(records=records, passed=False)
    finally:
        try:
            await adapter.disconnect()
        except Exception:
            pass

    records.append(
        {
            "type": "assertion",
            "candidate": CANDIDATE,
            "scenario": SCENARIO,
            "name": "all_targets_resolved",
            "passed": True,
            "hardGate": True,
        }
    )
    return RunResult(records=records, passed=True)


def required(name: str) -> str:
    value = os.environ.get(name, "").strip()
    if not value:
        raise RuntimeError(f"{name} wajib diisi untuk behavior test akun uji.")
    return value


async def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.parse_args(argv)
    try:
        config = SessionConfig(
            api_id=int(required("TELEGRAM_TEST_API_ID")),
            api_hash=required("TELEGRAM_TEST_API_HASH"),
            session=required("TELEGRAM_TEST_SESSION"),
        )
        targets = {role: required(name) for role, name in TARGET_ROLES.items()}
        result = await run_resolve_targets(lambda: create_telethon_adapter(config), targets)
    except Exception as error:
        result = RunResult(
            records=[
                metadata(),
                {
                    "type": "assertion",
                    "candidate": CANDIDATE,
                    "scenario": SCENARIO,
                    "name": "all_targets_resolved",
                    "passed": False,
                    "hardGate": True,
                    "code": failure_code(error),
                },
            ],
            passed=False,
        )

    for record in result.records:
        print(json.dumps(record, separators=(",", ":")))
    return 0 if result.passed else 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main(sys.argv[1:])))

File: spikes/telegram-engine/adapters/teleproto/behavior-resolve-targets.mjs
/** Resolve controlled Telegram targets without join/send side effects. */

import { performance } from "node:perf_hooks";
import { pathToFileURL } from "node:url";
import { SessionConfig, createTeleprotoAdapter } from "./adapter.mjs";

const SCENARIO = "resolve_controlled_targets";
const CANDIDATE = "teleproto";
const ADAPTER_VERSION = "teleproto@1.228.5";
export const TARGET_ROLES = Object.freeze({
  public_group: "TELEGRAM_TEST_PUBLIC_TARGET",
  discussion_channel: "TELEGRAM_TEST_DISCUSSION_CHANNEL",
});

function metadata() {
  return {
    type: "metadata",
    candidate: CANDIDATE,
    runtime: "node",
    runtimeVersion: process.version,
    adapterVersion: ADAPTER_VERSION,
    scenarioSet: "behavior-resolve-v1",
  };
}

function failureCode(error) {
  return typeof error?.code === "string" && error.code ? error.code : "TARGET_RESOLVE_FAILED";
}

export async function runResolveTargets({ createAdapter, targets, now = () => performance.now() }) {
  const records = [metadata()];
  const adapter = createAdapter();
  try {
    await adapter.connect();
    for (const role of Object.keys(TARGET_ROLES)) {
      const target = targets?.[role]?.trim();
      if (!target) throw new Error(`${TARGET_ROLES[role]} wajib diisi untuk behavior test.`);
      const started = now();
      const resolved = await adapter.resolveTarget(target);
      records.push({
        type: "sample",
        candidate: CANDIDATE,
        scenario: SCENARIO,
        sessions: 1,
        metric: "resolve_latency",
        value: Math.max(0, now() - started),
        unit: "ms",
        targetRole: role,
        entityType: resolved.entityType,
      });
      records.push({
        type: "assertion",
        candidate: CANDIDATE,
        scenario: SCENARIO,
        name: `${role}_resolved`,
        passed: true,
        hardGate: true,
        targetRole: role,
        entityType: resolved.entityType,
      });
    }
  } catch (error) {
    records.push({
      type: "assertion",
      candidate: CANDIDATE,
      scenario: SCENARIO,
      name: "all_targets_resolved",
      passed: false,
      hardGate: true,
      code: failureCode(error),
    });
    return { records, passed: false };
  } finally {
    try {
      await adapter.disconnect();
    } catch {
      // Preserve the original assertion. Resolve runner performs no external writes.
    }
  }

  records.push({
    type: "assertion",
    candidate: CANDIDATE,
    scenario: SCENARIO,
    name: "all_targets_resolved",
    passed: true,
    hardGate: true,
  });
  return { records, passed: true };
}

function required(name) {
  const value = process.env[name]?.trim();
  if (!value) throw new Error(`${name} wajib diisi untuk behavior test akun uji.`);
  return value;
}

async function main() {
  let result;
  try {
    const config = new SessionConfig({
      apiId: Number(required("TELEGRAM_TEST_API_ID")),
      apiHash: required("TELEGRAM_TEST_API_HASH"),
      session: required("TELEGRAM_TEST_SESSION"),
    });
    const targets = Object.fromEntries(Object.entries(TARGET_ROLES).map(([role, name]) => [role, required(name)]));
    result = await runResolveTargets({ createAdapter: () => createTeleprotoAdapter(config), targets });
  } catch (error) {
    result = {
      records: [
        metadata(),
        {
          type: "assertion",
          candidate: CANDIDATE,
          scenario: SCENARIO,
          name: "all_targets_resolved",
          passed: false,
          hardGate: true,
          code: failureCode(error),
        },
      ],
      passed: false,
    };
  }
  for (const record of result.records) process.stdout.write(`${JSON.stringify(record)}\n`);
  process.exitCode = result.passed ? 0 : 1;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) await main();
